### PR TITLE
ID-294 Initial work on more accessible sub-menus

### PR DESCRIPTION
**Pull request by lol768**
This adds:

* Left/right to navigate between top-level items or adjacent dropdowns
* Down to open the sub-menu, auto-focus on first item
  * Uses Bootstrap's plugin function, which adds `aria-expanded` for us and provides navigation through sub-items with <kbd>Up</kbd>/<kbd>Down</kbd>
  * Patches Bootstrap's plugin to provide:
    * Wraparound at bottom/top of list
    * Jump to item with first letter press
* Up to close the sub-menu

This PR also, via CSS, adds a caret automatically to all dropdown links so we have a visual indication that you can hit down/enter to open the menu. This may well cause problems with sites which&mdash;in contrast with Sitebuilder&mdash;include these already (think Webgroups). Any thoughts on this?

Example of it in use:

output

### CODE REVIEW

**Comment by lol768** [comment]
Collapsing the dropdown works now. Do we want <kbd>Left</kbd> to close an open dropdown? What should <kbd>Up</kbd> on the first item of a dropdown do?

Do we support nested dropdown/sub-menus?

**Comment by halfninja** [comment]
I think there's [guidance ](https://www.w3.org/WAI/GL/wiki/Using_ARIA_menus#Keyboard_Actions_if_focus_is_on_a_menu_item) in the accessibility specs that say to make them behave as an application's menu bar would - so that would mean when you have a menu open, left and right jump between menus on that bar. But I wouldn't say it's essential, so if it's a pain don't worry.

File: js/navigation.jquery.js [hunk 1]
```diff
@@ -304,6 +304,90 @@ class Navigation {
...
+
+    // Accessibility
+    this.$container.on('keydown', (ev) => {
+      const $focus = $(document.activeElement);
+
+      const $li = $focus.parent();
+      const isEnterOrDown = ev.keyCode === 40 || ev.keyCode === 13;
+      const isEnterOrUp = ev.keyCode === 38 || ev.keyCode === 13;
+
+      // Is there a Bootstrap method to test this?
```
Comment (matmannion) [inline js/navigation.jquery.js]: Could listen for `shown.bs.dropdown` and `hidden.bs.dropdown` but your way is probably fine

File: js/navigation.jquery.js [hunk 2]
```diff
@@ -304,6 +304,90 @@ class Navigation {
     // Handle in-page bookmarks.
     if (document.location.hash) this.hashChanged();
     $(window).on('hashchange', $.proxy(this.hashChanged, this));
+
+    // Accessibility
+    this.$container.on('keydown', (ev) => {
```
Comment (matmannion) [inline js/navigation.jquery.js]: Can we have this behaviour be optional but enabled by default?

File: js/navigation.jquery.js [hunk 3]
```diff
@@ -304,6 +304,90 @@ class Navigation {
...
+      const $focus = $(document.activeElement);
+
+      const $li = $focus.parent();
+      const isEnterOrDown = ev.keyCode === 40 || ev.keyCode === 13;
```
Comment (matmannion) [inline js/navigation.jquery.js]: Can we extract these keycodes to a single place so it's clear which ones we're catching at a glance (enter/up/down/left/right I think)

Comment (lol768) [reply js/navigation.jquery.js]: We could do that, or I could change to use <a href="https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key">`event.key`</a> - do you think browser support is good enough?

File: js/navigation.jquery.js (hunk 2 above)
Comment (lol768) [reply js/navigation.jquery.js]: How do you want the opt out to work, `data-` attribute on the `nav` element or something?

Comment (matmannion) [reply js/navigation.jquery.js]: data on .id7-navigation, which is how the other optional functionality works (e.g. fixedHeader, fixedNav etc)

File: js/navigation.jquery.js (hunk 3 above)
Comment (matmannion) [reply js/navigation.jquery.js]: Yes, I think support's good enough.

**Comment by lol768** [comment]
I'm looking into a blocking bug that prevents clicks on the dropdown items from working after the menu has been opened using the keyboard

**Comment by lol768** [comment]
Also fixes dropdown active styling:

image

**Comment by lol768** [comment]
Mat pointed out that the dropdown behaviour stopped users being able to click/hit enter on the top-level nav items. We discussed why this was the case.

Unfortunately fixing this has been ~a complete pain in the ass due to how Bootstrap's dropdowns work~ more challenging than expected. However, with a bit of patching, it does now function as you'd expect. If you click the trigger button, or focus on it and hit enter, you'll navigate to the page you'd expect.

if we update Bootstrap we should check this patch... Ideally in ID8 we'd have either a split button or just use a sub-item for the top-level items. Then life would be easy, because click-to-open dropdown would be fine.

File: js/id7-bundle.js [hunk 4]
```diff
@@ -12,6 +12,85 @@ require('bootstrap/js/button');
 require('bootstrap/js/carousel');
 require('bootstrap/js/collapse');
 require('bootstrap/js/dropdown');
+
+// ID-294 We use non-standard dropdowns that need to have different click actions..
```
Comment (lol768) [inline js/id7-bundle.js]: Worst part of PR starts here..

File: less/navigation.less [hunk 5]
```diff
@@ -199,6 +199,18 @@
           display: none;
         }
       }
+
+      &.dropdown > a::after {
```
Comment (lol768) [inline less/navigation.less]: Cribbed from the `.caret` styles.. Maybe there's a mixin we can use

Comment (lol768) [reply less/navigation.less]: Because `.caret` is a class based set of rules, we can use it as a mixin.

File: js/navigation.jquery.js (hunk 2 above)
Comment (lol768) [reply js/navigation.jquery.js]: Done with `data-no-accessibility="true"`

**Comment by lol768** [comment]
Can I get a review on this following the reasonably big changes made to fix click support?

**Comment by lol768** [comment]
We now use `data-keyboard` per Slack.

**Pull request merged**